src/docopt/dispatch.nim

```diff
@@ -0,0 +1,59 @@
+import docopt, macros, strutils, sequtils
+
+proc to[T](v: Value): T =
+  when T is SomeInteger:
+    T(parseInt($v))
+  elif T is bool:
+    v.toBool
+  elif T is string:
+    $v
+  elif T is seq:
+    @v
+
+proc discardable(b: bool): bool {.discardable.} = b
+
+macro dispatchProc*(args: Table[string, Value], procedure: proc, conditions: varargs[string]): untyped =
+  var check = newLit(true)
+  for condition in conditions:
+    check = nnkInfix.newTree(newIdentNode("and"), check, quote do:
+      `args`[`condition`] == true)
+  let
+    procImpl = procedure.getImpl
+    argIt = newIdentNode("argIt")
+    setArguments = newIdentNode("setArguments")
+  if conditions.len == 0:
+    let name = $procImpl[0]
+    check = quote do:
+      `args`[`name`] == true
+  var
+    argVariables = nnkVarSection.newTree()
+    setArgs = nnkBracket.newTree()
+    findArgs = newStmtList()
+    call = nnkCall.newTree(procImpl[0])
+    i = 0
+  for arg in procImpl[3][1..^1]:
+    argVariables.add arg
+    for subArg in arg[0..^3]:
+      setArgs.add newLit(false)
+      let
+        name = subArg
+        nameStr = $subArg
+        kind = arg[^2]
+      findArgs.add quote do:
+        if `argIt`.strip(chars = {'<', '>', '-'}) == `nameStr`:
+          `setArguments`[`i`] = true
+          `name` = `args`[`argIt`].to[:`kind`]()
+      call.add name
+      inc i
+  argVariables.add nnkIdentDefs.newTree(setArguments, newEmptyNode(), setArgs)
+  result = quote do:
+    if `check`:
+      `argVariables`
+      for `argIt` in `args`.keys:
+        `findArgs`
+      if `setArguments`.allIt(it == true):
+        `call`
+      discardable true
+    else:
+      discardable false
+  echo result.repr
```

Comment:
Wait, is `echo` part of normal operation or is this debugging?

Reply:
No, that's debugging output.

Reply:
Yeah, that just prints out during compile-time what it turned the `dispatchProc` into. But with `--expandMacro` now being implemented in the compiler this should probably be removed.